### DISCUSSION TITLE
WPF : How to handle image types that don't render via BitmapImage?

### DISCUSSION THREAD

**Discussion by ChuckSavage** [dlemstra/Magick.NET#730] [Help]
I recently set my app to permit all file types supported by Magick.NET. The first image type that I've noticed that busted my app is the Photoshop PSD file.

In my ImageConverter class, I in brief do.

```
var image = new BitmapImage();
image.BeginInit();
image.UriSource = new Uri(path);
image.EndInit();
return image;
```

The PSD file exceptions on the `image.EndInit()` with `System.NotSupportedException: No imaging component suitable to complete this operation was found.`

The workaround I've found is to load the psd file and re-write a temp file with a .jpg extension, then display that.

Is there something more suitable than this?

Also, is there some internal Magick.NET solution that knows what isn't a BitmapImage? Or should I just catch that exception, generic as it is?

**Comment by ChuckSavage** [comment]
So far I've decided to go the second route, catch the exception - still I'd like to know if there is something WPF friendly within the library.

```
    catch (System.NotSupportedException ex)
    {
        if (ex.Message.StartsWith("No imaging component suitable"))
        {
            image = new BitmapImage();
            image.BeginInit();
            using (var img = new ImageMagick.MagickImage(path)
            {
                Func<int, FileInfo> fileInfo = reverseCount =>
                {
                    string name = "imageConverter" + (MaxWriteRetries - reverseCount).ToString();
                    return temp = Lib.TempFile(name).ChangeExtension("jpg");
                };
                // Sometimes there are errors on write, retry MaxWriteRetries times with different names
                UnableToOpenBlob.OpenBlob(null, reverseCount => img.Write(fileInfo(reverseCount)), MaxWriteRetries);
            }
            image.UriSource = new Uri(temp.FullName);
            image.CacheOption = BitmapCacheOption.OnLoad;
            image.CreateOptions = BitmapCreateOptions.IgnoreImageCache;
            image.EndInit();
        }
        else
            throw;
    }
```

**Comment by dlemstra** [comment]
There is a NuGet package called `Magick.NET.SystemWindowsMedia` that will give you a `magickImage.ToBitmapSource()` method that you could use to change a `MagickImage` into `BitmapSource` object.

**Comment by ChuckSavage** [comment]
Alright, cool. I was thinking there had to be something. So would it be considered a StreamSource? From my example above, would it become. Otherwise, I'm not sure how this works.

```
            image = new BitmapImage();
            image.BeginInit();
            using (var img = new ImageMagick.MagickImage(path)
            {
                image.StreamSource = img.ToBitmapSource();
            }
            image.EndInit();
```

**Comment by dlemstra** [comment]
The `ToBitmapSource` method returns a `BitmapSource` that is the base class of the `BitmapImage` class. Where you are using the `image` you can probably use the result of `img.ToBitmapSource()`.
